Dynamickademlia

svetomir opened the pull request:
Main goal of this PR is having Kademlia check against dynamic Bee cluster.
Also, several other features had to be developed before this kind of check was possible.
Therefore, new things introduced in this PR are listed bellow:
- Kadmelia check is rewritten to support dynamic cluster
- type Node is introduced in the pkg/bee to allow better code structure and management (Bee Client type is part of Node)
- type NodeGroup is now collection of Nodes (not Clients anymore); also, new methods are added to support stopping/starting/adding/removing nodes from the node group (and hence from the cluster)
- pkg/k8s/statefulset is improved to support stopping/starting/adding/removing nodes
- multiple options are added to the start cluster and start node commands to improve usability
- several functions are renamed due to code readability

There are some things in the Kadmelia check that seem redundant. Those things can be treated as legacy, and I left them in to keep support for cluster started by Helm. Changing this would require changing tests on all places in CI/CD. Once we migrate all tests to the new approach, those things can be rewritten and code will become cleaner.

NOTE: Kademlia check is failing during dynamic test. Maybe [check itself](https://github.com/ethersphere/beekeeper/blob/57434a06dc81396f1f309c75ef4c30ab55dfe623/pkg/check/kademlia/kademlia.go#L192) needs to be updated - I am guessing that conditions are not viable anymore. But, also we might have some issue with topology inside Bee.
Anyways, please try to run test to replicate the issue, and afterwards we should discuss about updating this check and do further investigation in Bee if needed.
Command for running test is:
`./dist/beekeeper check kademlia --dynamic --start-cluster -b 2 -c 15`

janos reviewed (commented):
Very nice. There is room to reduce the code by pulling some of the common functionalities but as we discussed, that should be left for the next iterations when other tests are updated to dynamic ones, resulting more clarity how to do these abstractions.

acud commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+
+		fmt.Println("Checking Kademlia")
+		if err := checkKademliaD(topologies); err != nil {
+			fmt.Printf("Kademlia check failed: %v\n", err)
```
should we not annotate and return the error here? or maybe we should have a feature toggle for this like with the go test runner? `--failfast`?

svetomir replied:
We should. But I left it like this so I could finish dynamic test as Kademlia check is failing and has to be redefined. 
Nevertheless, `--failfast` seems like a good feature and we can implement it for all checks.

acud commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -60,3 +188,32 @@ func Check(cluster *bee.Cluster) (err error) {
 	return
 }
+
+func checkKademliaD(topologies bee.ClusterTopologies) (err error) {
```
this never will return an error, since the allocation in `L202` will shadow the named return. Under all error conditions, the error will only be printed, but never returned.

svetomir replied:
Same reason as above. Kademlia check needs to be redefined.

acud commented on pkg/k8s/statefulset/client.go:
```diff
@@ -29,6 +30,53 @@ type Options struct {
+}
+
+// Scale scales StatefulSet
+func (c *Client) Scale(ctx context.Context, name, namespace string, replicas int32) (err error) {
```
using a named return does not give any benefit here

acud reviewed (commented):
nice work @svetomir.
I have a few general comments:
- There's a lot of type translation being done, which introduces duplicates of existing types which are already scattered around beekeeper, and I'm not sure what is the reason for that
- Errors annotation should be unified across the codebase. Some use `%s`, some `%w`, some `%v`. We should use `%w` everywhere
- Dynamic kademlia test should be in a separate file in the `kademlia` folder
- I would even go as far as separating it from the `kademlia` command too, just in the same way that the different checks for pinning are done in different commands. It would make the code in the command folder more readable
- I would personally abstract the actual handling of `kademlia.Actions` to something like `scenario.Step`, which is a more generalised approach in which other tests could make use of the same types and same underlying logic that actually does the whole k8s heavy lifting (i.e. start/stop/add/remove then execute a generic condition that must pass on every `Step`). Until we have this pulled out to a separate package, all new tests will have to either duplicate this logic or reference the kademlia package

svetomir commented on Makefile:
```diff
@@ -1,6 +1,6 @@
 GO ?= go
 GOLANGCI_LINT ?= golangci-lint
-GOLANGCI_LINT_VERSION ?= v1.24.0
+GOLANGCI_LINT_VERSION ?= v1.30.0
```
Thanks, I updated it.

svetomir commented on cmd/beekeeper/cmd/check_kademlia.go:
```diff
@@ -1,43 +1,215 @@
-			cluster := bee.NewCluster("bee", bee.ClusterOptions{
+			k8sClient, err := setK8SClient(c.config.GetString(optionNameKubeconfig), c.config.GetBool(optionNameInCluster))
+			if err != nil {
+				return fmt.Errorf("creating Kubernetes client: %v", err)
```
Fixed.

svetomir commented on cmd/beekeeper/cmd/check_kademlia.go:
```diff
@@ -1,43 +1,215 @@
+				}
+
+				if err := nGroup.Wait(); err != nil {
+					return fmt.Errorf("starting nodes: %v", err)
```
Fixed.

svetomir commented on cmd/beekeeper/cmd/check_kademlia.go:
```diff
@@ -1,43 +1,215 @@
+					for i := 0; i < bootnodeCount; i++ {
+						if err := bg.AddNode(fmt.Sprintf("bootnode-%d", i), bee.NodeOptions{}); err != nil {
+							return fmt.Errorf("adding bootnode-%d: %v", i, err)
+						}
```
Fixed.

svetomir commented on cmd/beekeeper/cmd/start_cluster.go:
```diff
@@ -58,54 +67,77 @@ func (c *command) initStartCluster() *cobra.Command {
+			}
+
+			if err := bnGroup.Wait(); err != nil {
+				return fmt.Errorf("starting bootnodes: %v", err)
```
Fixed.

svetomir commented on pkg/bee/nodegroup.go:
```diff
@@ -80,16 +117,20 @@ type NodeGroupAddresses map[string]Addresses
 	for _, m := range msgs {
 		if m.Error != nil {
-			return nil, fmt.Errorf("%s: %w", m.Name, m.Error)
+			return nil, fmt.Errorf("%s: %v", m.Name, m.Error)
```
Fixed.

svetomir commented on cmd/beekeeper/cmd/helpers.go:
```diff
@@ -8,8 +8,8 @@ import (
-// newBeeDefaultConfig returns default Bee node configuration
-func newBeeDefaultConfig() *k8sBee.Config {
+// defaultBeeConfig returns default Bee node configuration
+func defaultBeeConfig() *k8sBee.Config {
```
Renamed.

svetomir commented on cmd/beekeeper/cmd/helpers.go:
```diff
@@ -45,13 +45,13 @@ func newBeeDefaultConfig() *k8sBee.Config {
-// newDefaultNodeGroupOptions returns default node group options
-func newDefaultNodeGroupOptions() *bee.NodeGroupOptions {
+// defaultNodeGroupOptions returns default node group options
+func defaultNodeGroupOptions() *bee.NodeGroupOptions {
```
Renamed.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+	fmt.Println("Checking connectivity")
+	err = fullconnectivity.Check(ctx, cluster)
+	if err != nil {
+		fmt.Printf("Full connectivity not present\n")
```
Done.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+			fmt.Printf("node %s is added\n", nName)
+		}
+
+		time.Sleep(5 * time.Second)
```
Thanks, will be done.

svetomir commented on pkg/k8s/bee/node.go:
```diff
@@ -19,8 +19,117 @@ const (
+}
+
+// Delete deletes Bee node from the cluster
+func (c *Client) Delete(ctx context.Context, o DeleteOptions) (err error) {
```
Fixed.

svetomir commented on pkg/k8s/bee/node.go:
```diff
@@ -19,8 +19,117 @@ const (
+		return false, fmt.Errorf("statefulset %s in namespace %s ready replicas: %w", o.Name, o.Namespace, err)
+	}
+
+	if r != 1 {
+		return false, nil
+	}
+
+	return true, nil
```
Thanks, I updated it.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+// NewClient returns Bee client
+func NewClient(opts ClientOptions) (c *Client) {
+	c = &Client{
+		retry: 5,
```
Done.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+	fmt.Println("Checking connectivity")
+	err = fullconnectivity.Check(ctx, cluster)
+	if err != nil {
+		fmt.Printf("Full connectivity not present\n")
```
@janos Done.
@acud This was written in the issue for writing Kademlia check. Anyways, I mentioned we should have a meeting in which we could rethink the whole Kademlia check. So, lets do it that way, and than rewrite the check if needed.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+		retry: 5,
+	}
+
+	if opts.APIURL != nil {
```
Good point. But, lets do it through different PR as we can use this across whole Beekeeper codebase.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+}
+
+// Balance represents node's balance with peer
+type Balance struct {
```
All Beekeeper packages have their own types, independent from one another.
There was a need for Bee client package from the beginning of Beekeeper project, but this client was never written as part of Bee codebase. Therefore, Package beeclient is written in Beekeeper, with idea that one day it might become part Bee. The rest of Beekeeper code was designed to be completely independent from it, hence type translation.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+		return Balance{}, fmt.Errorf("get balance with node %s: %w", a.String(), err)
+	}
+
+	return Balance{
```
Same answer as above - type translation.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+}
+
+// Balances represents Balances's response
+type Balances struct {
```
Same answer as above - type translation.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+}
+
+// PinnedChunk represents pinned chunk
+type PinnedChunk struct {
```
Same answer as above - type translation.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+
+// PinnedChunks represents pinned chunks
+type PinnedChunks struct {
+	Chunks []PinnedChunk
```
Same answer as above - type translation.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+func (c *Client) Overlay(ctx context.Context) (o swarm.Address, err error) {
+	var a debugapi.Addresses
+	for r := 0; r < c.retry; r++ {
+		time.Sleep(2 * time.Duration(r) * time.Second)
```
Can you please write an example, as timeout here is generated automatically depending on r.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+			pingStream <- PingStreamMsg{
+				Node:  node,
+				RTT:   rtt,
+				Index: i,
```
This PR introduces **node** concept which makes Index not needed. But, removing it everywhere needs to be tested first.
I removed it in places I was able to test while I was writing this PR. It is left here (line 284) and in few other places as it needs to confirmed safe for removal.
I plan to restructure these things through separate PRs which would tackle legacy stuff and some technical debt I noticed.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+}
+
+// RemoveChunk removes the given chunk from the node's local store
+func (c *Client) RemoveChunk(ctx context.Context, chunk *Chunk) (err error) {
```
Same answer as above.

svetomir commented on pkg/bee/client.go:
```diff
@@ -0,0 +1,538 @@
+
+// UploadChunk uploads chunk to the node
+func (c *Client) UploadChunk(ctx context.Context, chunk *Chunk, o api.UploadOptions) (err error) {
+	p := bmtlegacy.NewTreePool(chunkHahser, swarm.Branches, bmtlegacy.PoolSize)
```
Same answer as above. Not part of this PR.

svetomir commented on pkg/bee/nodegroup.go:
```diff
@@ -394,20 +505,28 @@ func (g *NodeGroup) PeersStream(ctx context.Context) <-chan PeersStreamMsg {
-	delete(g.nodes, name)
+// NodeReady returns node's readiness
+func (g *NodeGroup) NodeReady(ctx context.Context, name string) (ok bool, err error) {
+	ok, err = g.k8s.Ready(ctx, k8sBee.ReadyOptions{
```
Done.

svetomir commented on pkg/bee/nodegroup.go:
```diff
@@ -540,10 +659,68 @@ func (g *NodeGroup) StartNode(ctx context.Context, o StartNodeOptions) (err erro
+		}
+
+		fmt.Printf("%s is not ready yet\n", name)
+		time.Sleep(3 * time.Second)
```
Done.

svetomir commented on pkg/bee/nodegroup.go:
```diff
@@ -540,10 +659,68 @@ func (g *NodeGroup) StartNode(ctx context.Context, o StartNodeOptions) (err erro
+
+// StopNode stops node by scaling down its statefulset to 0
+func (g *NodeGroup) StopNode(ctx context.Context, name string) (err error) {
+	if err := g.k8s.Stop(ctx, k8sBee.StopOptions{
```
Done.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+	return
+}
+
+// Actions ...
```
Will be part of another pkg in the future.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+}
+
+// Actions ...
+type Actions struct {
```
Good point.
It is planned to have this in another pkg in the future. But, once we have more examples with other tests so we can design proper dynamic behaviour.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+}
+
+// CheckDynamic executes Kademlia topology check on dynamic cluster
+func CheckDynamic(ctx context.Context, cluster *bee.Cluster, o Options) (err error) {
```
Done.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -6,33 +6,161 @@ import (
+	fmt.Printf("Seed: %d\n", o.Seed)
+
+	fmt.Println("Checking connectivity")
+	err = fullconnectivity.Check(ctx, cluster)
```
Same answer as above about defining Kademlia check.

svetomir commented on pkg/check/kademlia/kademlia.go:
```diff
@@ -60,3 +188,32 @@ func Check(cluster *bee.Cluster) (err error) {
+					fmt.Printf("Error: %v\n", errKadmeliaBinConnected.Error())
+				}
+
+				if binDepth >= t.Depth && len(b.DisconnectedPeers) > 0 {
```
Agree. I already wrote about those things in general in my comment for this PR.
Let us have a meeting about possible changes to Kademlia check.

svetomir commented:
> nice work @svetomir.
> I have a few general comments:
> 
> * There's a lot of type translation being done, which introduces duplicates of existing types which are already scattered around beekeeper, and I'm not sure what is the reason for that
> * Errors annotation should be unified across the codebase. Some use `%s`, some `%w`, some `%v`. We should use `%w` everywhere
> * Dynamic kademlia test should be in a separate file in the `kademlia` folder
> * I would even go as far as separating it from the `kademlia` command too, just in the same way that the different checks for pinning are done in different commands. It would make the code in the command folder more readable
> * I would personally abstract the actual handling of `kademlia.Actions` to something like `scenario.Step`, which is a more generalised approach in which other tests could make use of the same types and same underlying logic that actually does the whole k8s heavy lifting (i.e. start/stop/add/remove then execute a generic condition that must pass on every `Step`). Until we have this pulled out to a separate package, all new tests will have to either duplicate this logic or reference the kademlia package

- You are right about it. Problem complexity and pkg design lead to this. If time allows us, we should discuss whether and how this should be mitigated.
- I fixed all of these things I found across whole code base.  
- I moved it into separate file as it makes more sense now. But, as static test is just a subset of a dynamic test, once we have proper pkg for dynamic actions this will all become just one method with options.
- I think we should redesign our CLI structure. This can be done at that time.
- Good point. Already planned it in one of the future PRs.

svetomir commented:
@acud @janos Thank you both for detailed reviews, they were very helpful.
I pushed changes that are in correlation with your reviews. Please check my commits and answers.
I also made a TODO list for future PRs which we'll discuss in the next meeting about Beekeeper.

janos reviewed (approved):
👍

acud reviewed (approved):
Thanks for addressing my comments :pray: Looking forward to seeing this in action
